### ISSUE TITLE
[UX] Support "multiple" HTML5 attribute on File elements

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#440]
The HTML5 "multiple" attribute allows multiple files to be selected and uploaded at once. In Backdrop, we should integrate this into the "managed_file" element and the File module field.

Sub-issue of https://github.com/backdrop/backdrop-issues/issues/200

This should be relatively portable from Drupal 8.

Commit: http://drupalcode.org/project/drupal.git/commit/8fd2b4779ec83c3e44a8caac65300b96e3a52ab8
Issue: http://drupal.org/node/625958

---
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/2430~~
~~PR by @quicksketch (based on @docwilmot's): https://github.com/backdrop/backdrop/pull/3030~~
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3755~~
PR by @quicksketch: https://github.com/backdrop/backdrop/pull/5125 (again based on @docwilmot's)

**Comment by quicksketch** [comment]
I'd love to get this in the next release, as we don't have time for this in 1.1.0.

**Comment by docwilmot** [comment]
This has been languishing for a year now. I've had a look at whats involved and its a bit above me. Can a person who is comfortable with File module and fields have a stab at this to move it along?

Just shepherding 1.3.0 issues.

**Comment by jenlampton** [comment]
Not sure this is going to make 1.3, but I'd also love to see it get in.

**Comment by klonos** [comment]
I have forgotten about this issue here. So forgive me if #1380 is a dupe. I had more in mind the D8 feature parity meta issue rather than the HTML5 meta.

**Comment by serundeputy** [comment]
Moving to 1.5; still a good idea!

**Comment by klonos** [comment]
I think that this one is pretty much what was left from #200. Do we have time to get this done for 1.5.0??

**Comment by olafgrabienski** [comment]
Getting it in 1.5 would be great! (In my case an important feature to be able to build a portfolio website for a photographer.)

**Comment by jenlampton** [comment]
We're less than one week away from release and there's no PR here yet. I'd say no, this likely won't make it. Officially moving to 1.x-future.

**Comment by olafgrabienski** [comment]
Unfortunately, I can't help with the code, but let us not forget this issue. (Just came back to it because I'm currently building a site where I could need it multiple image upload for a gallery.)

**Comment by olafgrabienski** [comment]
Added "[UX]" to the issue title, hope that's okay!

**Comment by laryn** [comment]
Everybody may be aware of this module already but I'll add it here since it hasn't been mentioned yet. Not sure if any code from this module can make this task easier:

https://backdropcms.org/project/html5_upload

**Comment by klonos** [comment]
These too:

https://www.drupal.org/project/multiupload_filefield_widget
https://www.drupal.org/project/multiupload_imagefield_widget

**Comment by docwilmot** [comment]
So I wanted to do this because it came up recently, and because @quicksketch said it "should be relatively portable from Drupal 8". I realise now that his "relatively" may have a different meaning. FieldAPI is such a morass of hooks and forms, I cant fix image defaults, which should arrays with this PR. Obviously this isn't going to make it to 1.12, so putting it back to rest till next time.

I think it may be close, the multiple-ness of fields is working, just need someone with FieldAPI skills to finish.

**Comment by quicksketch** [comment]
Hey @docwilmot, it's been a year since this work but I have to say: this is great! I tried out your PR and it actually nearly works! Very cool!

In the Drupal 8 approach, the `file_save_upload()` function was reworked to _always_ return an array instead of a single file. From an API-compability perspective, I don't think that's an option for us. Similar changes to the `managed_file` element (where `$element['fid']` is changed to `$element['fids']`) are also going to cause problems for contrib modules.

I took a stab at trying to maintain compatibility while also adding multiple file support. It's coming along but it's not fully working yet. In-progress PR at https://github.com/backdrop/backdrop/pull/3030

**Comment by klonos** [comment]
I have done a quick CR (mainly suggestions on inline comments, and user-facing text), and I have tested on the PR sandbox. Looks very promising! ❤️ ...thanks @docwilmot and @quicksketch 👍

**Comment by quicksketch** [comment]
Not sure how this got in 1.15.1. It should be backwards compatible when finished, but it's a pretty big change (and not a bug fix). I'd be happy to have it in 1.16.0 though. :smile:

**Comment by quicksketch** [comment]
@docwilmot refreshed his PR at https://github.com/backdrop/backdrop/pull/3755. I gave it a code review and it's looking great. Still lots of test failures related to File module.

**Comment by docwilmot** [comment]
Shout out to @hosef who I threatened to pass this on to. He didnt disagree, as far as I recall.

**Comment by BWPanda** [comment]
> I've had a look at whats involved and its a bit above me.

That's disconcerting... If @docwilmot is struggling with this, how can mere mortals hope to contribute to this issue? 🙂

**Comment by docwilmot** [comment]
I did attempt to pass this on to the Eternal @hosef but he did not heed my supplication. Perhaps I should sacrifice a goat. 🤔

**Comment by klonos** [comment]
> In the Drupal 8 approach, the `file_save_upload()` function was reworked to _always_ return an array instead of a single file. From an API-compability perspective, I don't think that's an option for us.

I understand how changing the return type of `file_save_upload()` would be an BC-breaking API change, but could/should we introduce a new `file_save_upload_multiple()` function, which always returns either an array of File objects, or `FELSE`? (when no file was uploaded)

**Comment by quicksketch** [comment]
> but could/should we introduce a new `file_save_upload_multiple()` function.

Yes, I think that may be a great approach and limit the complexity inside of `file_save_upload()`. We could potentially extract the similarities into a 3rd function and then put only the differences in `file_save_upload()` vs `file_save_upload_multiple()`.

**Comment by jenlampton** [comment]
Would (or could) this change also allow multiple files at a time for image fields?

**Comment by olafgrabienski** [comment]
> Would (or could) this change also allow multiple files at a time for image fields?

I think so, see my sandbox testing from October 2021:

> In the PR's sandbox website multiple files (e.g. images for Posts) are however already working nicely!

**Comment by stpaultim** [comment]
Just noting that this issue is still top of our forum wishlist. 
https://forum.backdropcms.org/feature-requests

I see that there is a PR, but re-reading the discussion it's clear to me exactly how close to completion this issue is or if there have been any new developments which might effect it's feasibility. 

Back in Oct 2021, @quicksketch wrote "I gave it a code review and it's looking great. Still lots of test failures related to File module."

**Comment by herbdool** [comment]
I attempted to resolve conflicts and bring the PR up to date. I tested it in Tugboat and it only sort of works - uploads multiple images on an image field but fatal error on saving the node. Not sure if that was how it was before.

**Comment by quicksketch** [comment]
This would be amazing to pick up!

**Comment by docwilmot** [comment]
This is down to 3 failing tests only. So some progress!

**Comment by olafgrabienski** [comment]
Great to see progress here! Already makes sense to test the PR?

I've had anyways a quick look in the sandbox: Uploading multiple images in Post works fine.
Uploading multiple PDF files at `admin/content/files` gave two errors (see below). The files have been uploaded, but only with temporary status.

The messages from DBlog:

- Warning: foreach() argument must be of type array|object, null given in file_add_form_submit() (line 350 of /var/lib/tugboat/core/modules/file/file.pages.inc).
- Warning: Undefined array key "upload" in file_add_form_submit() (line 350 of /var/lib/tugboat/core/modules/file/file.pages.inc).

**Comment by quicksketch** [comment]
@docwilmot I'm taking a look at your PR and will push some updates. A few general notes:

- It doesn't seem the update hook in `file.install` is necessary because default images are still single-value only.
- There are a couple places like `#file_value_callbacks` where previously a single FID was passed but now it's _always_ an array (even if its a single file upload). I think these should continue to pass either a single FID (if single upload) or multiple if multi-upload, similar to the way `file_save_upload()` is working.
- I'm trying to figure out why every row of a multi-value file field is returning an array of FIDs with only a single value.
- It looks like test failures are all around anonymous file uploading. I'm also trying to take a look at fixing those test failures.

**Comment by docwilmot** [comment]
I'd be happy to see this fixed. It's been beating me for too long now.

**Comment by quicksketch** [comment]
For now I filed a new PR at https://github.com/backdrop/backdrop/pull/5125, still heavily based on @docwilmot's work; I added a commit that uses a single FID in more place and fixes the issue @olafgrabienski reported about the `/file/add` form not saving properly. Still only a handful of test failures (same as before) so I think the changes are mostly positive.

**Comment by olafgrabienski** [comment]
@quicksketch – Thanks a lot for the new PR. It continues to work for multiple images in a post, but I'm still not able to upload two files without issues using the `file/add` form. When I upload two files and hit "Next" in the sandbox, I see those two messages above the form:

- Warning: Undefined array key "upload" in file_add_form_submit() (line 350 of /var/lib/tugboat/core/modules/file/file.pages.inc).
- Warning: foreach() argument must be of type array|object, null given in file_add_form_submit() (line 350 of /var/lib/tugboat/core/modules/file/file.pages.inc).

When I hit "Next" once more, the message is "You need to upload a file". When I cancel, I see that the files have been uploaded, but their status is "Temporary".

PS to be clear: uploading _one_ file via `file/add` works fine, but more than one not.

**Comment by quicksketch** [comment]
Thanks @olafgrabienski, I had multi-file upload working at one point at `file/add` but I must have broke it again. I appreciate you testing!

**Comment by quicksketch** [comment]
I'm still working on my PR. I realize I created some regressions from @docwilmot's earlier versions. I did not understand why the hidden `<input name="fid">` element was taking a comma-separated list of FID values. I switched its `#value` property to be an array, which obviously does not work since it needs to be a string when it's printed as an HTML element `value` attribute. I'm working on re-incorporating the previous approach.

**Comment by quicksketch** [comment]
I took another stab at this yesterday and updated https://github.com/backdrop/backdrop/pull/5125 with another commit that gets us closer to everything working. We're down to just a couple test failures.

**Comment by quicksketch** [comment]
Pushed another commit to https://github.com/backdrop/backdrop/pull/5125 that gets all tests running with one set of exceptions.

The last problem seems to be in `FileFieldValidateTestCase::testRequired()`. I'm at my wits end trying to figure out where the exceptions are coming from though.

As far as I can tell, what's happening is that somewhere there is a call to `file_load($fid)` or `file_load_multiple($fids)` where the File ID is incorrectly nested. Either: `file_load(array(4))` or `file_load_multiple(array(array(4)))`, where an array is being passed in instead of the single File ID. This then causes `entity_load_multiple()` to throw an exception, `array_flip(): Can only flip string and integer values, entry skipped`.

But making the whole thing a lot more frustrating: this only seems to happen on the GitHub runners. On my local, there are no exceptions and all tests pass.

**Comment by olafgrabienski** [comment]
@quicksketch – I've had a look at the sandbox of the updated PR.

Uploading multiple files via a node form works smoothly as before. Doing so at `file/add` now works generally as well. There is still one issue, though, using `file/add`: From a group of multiple files uploaded together, only the first gets the status "Permanent" (see screenshot below).

<img width="800" alt="Image" src="https://github.com/user-attachments/assets/9f67f273-d05b-4306-8f0d-87aaba27a7e8" />

**Comment by quicksketch** [comment]
Nice find @olafgrabienski! I'll need to add test coverage for that situation.

**Comment by indigoxela** [comment]
Is there anything I could help with? There seems to be some struggle with FileFieldValidateTestCase.

**Comment by quicksketch** [comment]
@indigoxela Yes, that would be wonderful. I've really been struggling with that one test failure. The weird thing is when I think I've cornered the problem with debugging lines, it stops happening. Maybe checkout my branch and make another PR for testing?

A nice side-effect of all this test debugging is I think I may have figured out our most common random test failure problem: https://github.com/backdrop/backdrop-issues/issues/7086.

**Comment by indigoxela** [comment]
> Maybe checkout my branch and make another PR for testing?

Should I use the one with the debugging stuff included? (debug_backtrace, print_r...)

I thought, I'd start with just running some tests locally, as my setup might be closer to how our GHA runner's set up. I might manage to get it failing.
Am I getting it right, that on your dev environment the test just passes?

**Comment by quicksketch** [comment]
> Should I use the one with the debugging stuff included? (debug_backtrace, print_r...)

Yeah, just roll-back the debugging commits with `git reset HEAD~1` (that pops off the last commit).

>  Am I getting it right, that on your dev environment the test just passes?

Yes, that's right. My local passes 100% of the time. Fortunately GHA _fails_ 100% of the time, so that at least is something to make things a little easier.

**Comment by indigoxela** [comment]
OK, "good" news, the test fails reliably here with 4 exceptions:

```
EXCEPTION: array_flip(): Can only flip string and integer values, entry skipped	Warning	entity.controller.inc	268	DefaultEntityController->load()	Exception
EXCEPTION: Array to string conversion	Warning	database.inc	2395	DatabaseStatementBase->execute()	Exception
EXCEPTION: Array to string conversion	Warning	cache.inc	488	BackdropDatabaseCache->getMultiple()	Exception
EXCEPTION: Array to string conversion	Warning	entity.controller.inc	324	DefaultEntityController->load()	Exception
```
Are these the "right" ones?

Edit: only running this one test on PHP 8.5.

**Comment by indigoxela** [comment]
Some quick debugging shows, that at that point in entity.controller.inc the array gets nested.

Instead of an expected structure like:
```
array (
  0 => '7',
)
```
...it's all of a sudden:
```
array (
  0 => 
  array (
    0 => 5,
  ),
)
```
Not sure yet, if that helps.

**Comment by quicksketch** [comment]
@indigoxela You're on exactly the right path. All tests are passing except the 4 exceptions in `FileFieldValidateTestCase`. And the problem seems to be an extra level of nesting when `entity_load_multiple()` is called.

> OK, "good" news, the test fails reliably here with 4 exceptions:

So you can get the exceptions locally? That's great news!

**Comment by indigoxela** [comment]
@quicksketch could you rebase (rebase of your own branch, not just by PR sync button) your branch and remove the debugging code and unrelated changes? It's all a bit messy now and I'm having trouble to get it working locally.

**Comment by quicksketch** [comment]
@indigoxela The PR at https://github.com/backdrop/backdrop/pull/5125 should now cleaned up.

**Comment by indigoxela** [comment]
I have to give up here. The test still reliably fails - no idea, why not in your dev environment @quicksketch. I thought it has something to do with function file_managed_file_process() but got stuck...

At some point an array gets wrapped in another array, so entity_load_multiple() chokes on it. But that was already clear.
I can't provide any additional clues. I don't even get how Simpletest handles file uploads (some curl magic). 😉 

This is the backtrace of functions from entity.controller.inc (not that I think, this provides much value):

```
array (
  0 => 'load',
  1 => 'entity_load_multiple',
  2 => 'entity_load',
  3 => 'file_load_multiple',
  4 => 'file_load',
  5 => 'file_field_update',
  6 => '_field_invoke',
  7 => 'field_attach_update',
  8 => 'invokeHook',
  9 => 'invokeHook',
  10 => 'save',
  11 => 'save',
  12 => 'node_form_submit',
  13 => 'form_execute_handlers',
  14 => 'backdrop_process_form',
  15 => 'backdrop_build_form',
  16 => 'backdrop_get_form',
  17 => 'node_page_edit',
  18 => 'system_block_view',
  19 => 'module_invoke',
  20 => 'buildBlock',
  21 => 'getContent',
  22 => 'renderBlock',
  23 => 'renderBlocks',
  24 => 'renderLayout',
  25 => 'render',
  26 => 'layout_route_handler',
  27 => 'menu_execute_active_handler',
)
```

One thing baffles me: the code change is still quite big, even after cleanup. With significant changes in central functions like file_save_upload(). Making sure this all is backwards compatible seems a challenge.

On the other hand there's a fully functional [contrib module](https://github.com/backdrop-contrib/html5_upload) that accomplishes the field part with less than 100 lines of code.
My impression is, that the effort here is disproportionate compared to the gain.

**Comment by docwilmot** [comment]
Replacing in `file_field_update()` line 256:
```
  if (!empty($entity->revision)) {
    foreach ($items as $item) {
      if ($file = file_load($item['fid'])) {
        file_usage_add($file, 'file', $entity_type, $id);
      }
    }
    return;
  }

```
with 
```
  if (!empty($entity->revision)) {
    foreach ($items as $item) {
      $fid = is_array($item['fid']) ? reset($item['fid']) : $item['fid'];
      if ($file = file_load($fid)) {
        file_usage_add($file, 'file', $entity_type, $id);
      }
    }
    return;
  }
```
that fixes the exceptions for me, but haven;t checked why `$item['fid']` is only array sometimes.

**Comment by quicksketch** [comment]
@docwilmot Yes! That is it! I can work from that. I pushed up a temporary commit and confirmed that the exception is fixed! I will see if I can handle this comprehensively. If possible, we'll want `$item['fid']` to stay an integer for contrib/custom modules.

@indigoxela I think the difference in complexity here can be explained. The [html5_upload](https://github.com/backdrop-contrib/html5_upload/tree/1.x-1.x) module only handles the File/Image widget when used as a Field module field. The work here on the other hand makes it so that the Form API element `$element['#type'] = 'file'` can handle multiple file uploads. That in turns makes it so that non-field situations (like `/file/add`) also support multiple file uploads. The current PR adds about 400 lines of new code, with about 200 lines of added code being new test coverage.

That said, it is changing a lot of fundamental file handling. My expectation (which may be incorrect) is that most backwards-compatibility will not be an issue because the `#multiple` Form API property is entirely new. Form API file fields simply did not support multiple files previously. I want to figure out what's going on with Field API making `$item['fid']` an array, since that does sound like a potential API break.

Thank you thank you to both @indigoxela and @docwilmot for helping overcome this technical hurdle. 😄

**Comment by indigoxela** [comment]
@docwilmot thank you so much for your help! Yes, I can confirm that this fixes the test. 🙏

But the PR still needs work, of course. Besides the array/string problem with `$item['fid']` the problem reported previously (only the first one marked "permanent" when uploading via file/add) still exists.

@quicksketch I am aware of the difference to the contrib solution. Yes, this PR does a lot more. 😉 The difference for users isn't that big, though. They can upload multiple files at once not only in fields with a specific widget. But multiupload in fields is IMO the main feature people need. Note: I don't oppose here in any way, I still support it. But sometimes I question the effort... 🙈

**Comment by quicksketch** [comment]
I found the source of the problem and why it happens during the `FileFieldValidateTestCase`. To reproduce:

- Create a multi-value file field that allows certain extensions, i.e. "jpg" only
- Then visit `/node/add/page` (or whatever type).
- Select multiple files, i.e. `a.jpg`, `b.jpg`, and `c.png`. Make sure the invalid extension is the last item selected alphabetically.
- Try to upload. The result is that `a.jpg` and `b.jpg` are uploaded, but incorrectly. A validation error occurs on `c.png`. You get something like this:
<img width="1177" height="673" alt="Image" src="https://github.com/user-attachments/assets/0b761ab4-a308-4f08-ba8d-09839e696dc3" />
The two valid files were uploaded but the last row and form handling is mangled. Saving the node results in `file_field_update()` getting passed an array of FIDs for the newly added item. In this example. there should be 6 values in the `$items` array, but there are only 5, and the last one contains two FIDs instead of one.

**Comment by indigoxela** [comment]
Seems like this is an additional problem, but the Exceptions in FileFieldValidateTestCase are back.

**Comment by quicksketch** [comment]
Indeed, there is yet another problem. I can now reproduce that problem locally as well. The trick is that JavaScript needs to be disabled so that the file can to be uploaded the same time as saving the node.

- Set up a multi-value file field
- Visit `/node/add/page`
- Disable JavaScript in the browser
- Select one or more files, the quantity does not matter here
- Do not click the "Upload" button (that would work fine)
- Scroll down and save the node so that the file is uploaded in the same request as the node is saved

This situation rarely occurs in real-world usage, but I'll see if this can be fixed.

**Comment by quicksketch** [comment]
I have a better understanding now. The unexpected scenario in my previous comment occurs because the form is not rebuilt like it is if you click the "Upload" button (or the AJAX upload does it for you). When the form is rebuilt with new values, it does the expansion from one file field into multiple different field items.

If the node form is saved without uploading the files first, then the form save handler is executed, and the form is never rebuilt. This means that `file_field_insert()` and `file_field_update()` must handle the expansion themselves. I added a new function `file_field_expand_items()` to do this.

I was worried about contrib projects that might be implementing `hook_field_insert()` or `hook_field_update()`, but these aren't really "hooks"; they're callbacks. Only the defining module is called ("file" in this case). But for modules implementing `hook_entity_insert()` or `hook_entity_update()`, the `$items` parameter is modified by reference, so adding/expanding the items into different values will make things predictable/unchanged for those hooks.

I updated my PR. Now all tests and all situations I know of should be working. There is still more work to do in writing new test coverage:

- Ensure multiple files uploaded at `/file/add` are all marked permanent.
- Test the situation of uploading one invalid file amongst other valid files (the valid files should upload and the invalid one should not).
- We could use more test coverage of clicking the "Upload" button vs using "Save" on the node form (not directly related to this change).

@indigoxela @olafgrabienski @docwilmot I would appreciate further review and sandbox testing. Thank you!

**Comment by indigoxela** [comment]
Wow, with JS disabled and upload button not clicked! This is it!

However, the latest changes introduced a new problem, now file upload in field widgets doesn't work at all... 😬 Neither for a single value image field, nor for a multivalue file field. The whole field widget looks weird. Not even selecting an existing image from library still works.

Loads of other tests failing now.

The only upload that still (sort of) works is the one on file/add - but even that one causes `Warning: Undefined array key "fid" in file_managed_file_validate()`. All files are permanent after upload, though.

I'm guessing, the last commit broke things for field widgets.

EDIT: yes, reverting the last commit fixes file field widgets.

**Comment by quicksketch** [comment]
Oops, sorry. It _was_ all working and then I changed something. I'll fix it again tonight.

**Comment by indigoxela** [comment]
No need to excuse, your effort's highly appreciated. 🙏

**Comment by quicksketch** [comment]
Now it's fixed again. 🙂

**Comment by indigoxela** [comment]
Awesome, everything works now like a charm, file fields, image fields, dropping images into the editor, adding files via file/add 🎉 

Now for the tricky part. Existing contrib module [html5_upload](https://backdropcms.org/project/html5_upload), which currently shows a usage of 473.
That one breaks. No errors, but uploading no longer works for fields with a cardinality other than "1". Selecting files just ends up with "no files selected". 😬

As core provides everything this module provides, we could disable it in an update hook... 🤔

**Comment by olafgrabienski** [comment]
I've had a look in the sandbox, and everything works for me there as well, including status and use count on `admin/content/files`. Fantastic! (I didn't look at HTML5 Upload which can't be installed in the sandbox.)

**Comment by herbdool** [comment]
Looks to be a good candidate for 1.34.

**Comment by herbdool** [comment]
In my testing it works too. I couldn't tell if it still needed more automated tests or not; I'll leave that label.

The "needs work" label I think is just in relation to adding an update hook to disable `html5_upload`.

**Comment by herbdool** [comment]
I've added that update hook (to disable `html5_upload`) and pushed to @quicksketch PR.

**Comment by indigoxela** [comment]
"needs work" is also for the phpcs nagging still present in touched (moved) code. 😉

And disabling the contrib module was only an idea. Maybe @laryn (maintainer of the affected contrib module) has some thoughts to share here?

**Comment by indigoxela** [comment]
FTR: our phpcs ruleset got updated and that's now also considered here: https://github.com/backdrop/backdrop-issues/issues/7088 
(In case the trigger for the ruleset update was this issue @quicksketch )

**Comment by quicksketch** [comment]
After a bit more work on coding standards and updating the cspell rules in https://github.com/backdrop/backdrop-issues/issues/6280, the PR is now passing all linting. We still need the expanded automated test coverage but we're closing in here.

**Comment by laryn** [comment]
> Maybe laryn (maintainer of the affected contrib module) has some thoughts to share here?

I've always seen `HTML5 Upload` as a stopgap module and disabling it once core has fixed this more comprehensively makes total sense.

**Comment by quicksketch** [comment]
PR is now updated with test coverage for multiple file uploading. Simpletest's cURL-based handling did not yet have support for uploading multiple files at a time. Now it supports it. [cURL support for multiple files](https://www.php.net/manual/en/curlfile.construct.php#example-2) is actually pretty great and it was not too bad to adjust.

Test coverage includes `/file/add` uploading multiple files at a time, ensuring they are marked permanent. And validation of files when one file fails validation but other files pass; the acceptable files are uploaded while the rejected ones create a validation message. Again checking for permanent status on save.

These are the main scenarios I saw fail when writing the PR. Combined with existing tests passing I think this is sufficient coverage.

For transparency, I want to put it in writing that the issue we had for years was solved by Claude Code (Sonnet 4.6). The added test coverage was also assisted by the same model. This is the first core PR I've used AI in any significant way; even if the end result was only about 60 lines out of 1200.

**Comment by herbdool** [comment]
I've tested and read through all the code. It looks good to me! This has gotten a lot of eyes on it over the years.

I was trying to figure out what drupal patch/commit was the original inspiration. The original link not longer works. Maybe https://www.drupal.org/files/multiple_fapi_uploads_625958_157.patch? Though the drupal version didn't care about BC.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/5125 into 1.x for 1.34.0!! 🎉 🎉 💃 🥳 🎂 

10 years of work finally completed! Huge thank you to @docwilmot and @opi for the original work on the pull request(s). Thank you @herbdool, @indigoxela, and @olafgrabienski for extensive feedback and testing.
